### PR TITLE
evm-watcher: Add logic to group events for dispatch

### PR DESCRIPTION
**Pull request by ali-bahjati**
It combines multiple events to form a `UpdatePriceFeedsAggregateInfo` and passes it down to watchers. 

There is one assumption in the code which is hard to get around, and that is `updatePriceFeeds` is called only once in a transaction. I couldn't find an easy workaround to handle multiple of them but 
I will tackle it in the future if needed.

### CODE REVIEW

File: third_party/pyth/evm-watcher/src/listener.ts [hunk 1]
```diff
@@ -32,31 +43,135 @@ export class Listener {
...
+        return;
+      }
+
+      const updatePriceFeedsInfo: UpdatePriceFeedsInfo = {
+        batchUpdates: this.txBatches.get(txHash)!,
+        ...updateValues,
+      }
+
+      const waitForTimestampCondition = () => this.blockToTimestamp.has(event.blockNumber);
+      const waitForTimestampResult = await waitForCondition(waitForTimestampCondition, 5000);
```
Comment (jayantk) [inline third_party/pyth/evm-watcher/src/listener.ts:149]: i don't love this chained waiting logic but I can't think of a better way to do it, so I guess this is fine.

Comment (ali-bahjati) [reply third_party/pyth/evm-watcher/src/listener.ts:149]: Me neither, I will add an issue for the future to also solve bad assumption that I made. 

The solution I'm thinking about is parsing the transaction logs manually (with events abi) instead. This way no chaining is needed, but it's less clean as we need to match log to event ourselves.

File: third_party/pyth/evm-watcher/src/listener.ts [hunk 2]
```diff
@@ -3,18 +3,29 @@ import { Contract, EventData } from "web3-eth-contract";
...
+import { BatchPriceFeedInfo, BatchPriceFeedUpdateEventValues, PriceFeedInfo, UpdatePriceFeedsAggregateInfo, UpdatePriceFeedsEventValues, UpdatePriceFeedsInfo } from "./events";
 
 type ListenerConfig = {
   wsEndpoint: string,
   pythContract: string
 };
 
+type BatchHash = string;
+function getBatchHash(chainId: NumberInString, sequenceNumber: NumberInString): BatchHash {
+  return `${chainId}-${sequenceNumber}`;
```
[the remark this answers is not in the document]
Comment (ali-bahjati) [reply third_party/pyth/evm-watcher/src/listener.ts:19]: Sounds good. thanks!

File: third_party/pyth/evm-watcher/src/utils.ts [hunk 3]
```diff
@@ -1,7 +1,8 @@
 export type HexString = string;
-export type UnixTimestamp = number;
+export type UnixTimestampInString = string;
+export type NumberInString = string;
```
[the remark this answers is not in the document]
Comment (ali-bahjati) [reply third_party/pyth/evm-watcher/src/utils.ts:3]: Yeah you are right. Thank you!

File: third_party/pyth/evm-watcher/src/listener.ts [hunk 4]
```diff
@@ -32,31 +43,135 @@ export class Listener {
...
     var gasPrice: number;
 
     // In some networks such as BNB effective gas price is not provided in response.
     if (txReceipt.effectiveGasPrice !== undefined) {
-      gasPrice = txReceipt.effectiveGasPrice;
+      return txReceipt.effectiveGasPrice;
     } else {
-      const tx = await this.web3.eth.getTransaction(txhash);
-      gasPrice = Number(tx.gasPrice);
+      return undefined;
```
[the remark this answers is not in the document]
Comment (ali-bahjati) [reply third_party/pyth/evm-watcher/src/listener.ts:61]: you are right, it was something else before and I forget to refactor it. Thanks!

File: third_party/pyth/evm-watcher/src/listener.ts [hunk 5]
```diff
@@ -32,31 +43,135 @@ export class Listener {
...
-    return gasPrice * txReceipt.gasUsed / Math.pow(10, 18);
   }
 
   start() {
     this.web3.eth.subscribe('newBlockHeaders', (_error: Error, blockHeader) => {
-      console.dir(blockHeader);
+      this.blockToTimestamp.set(blockHeader.number, blockHeader.timestamp.toString());
+    });
+
+    const priceFeedUpdateEmitter = this.pythContract.events.PriceFeedUpdate(undefined, async (_error: Error, event: EventData) => {
```
[the remark this answers is not in the document]
Comment (ali-bahjati) [reply third_party/pyth/evm-watcher/src/listener.ts:70]: Yeah, makes sense.

**Pull request merged**